critwall/src/main/java/org/opencadc/critwall/BucketSelector.java

```diff
@@ -67,18 +67,88 @@
 
 package org.opencadc.critwall;
 
+import ca.nrc.cadc.util.StringUtil;
+import java.util.Iterator;
+import java.util.TreeSet;
 import org.apache.log4j.Logger;
+import org.opencadc.inventory.Artifact;
+import org.opencadc.inventory.InventoryUtil;
+
 
-/**
- *
- * @author pdowler
- */
 public class BucketSelector {
-    private final Logger log = Logger.getLogger(BucketSelector.class);
-    public final String bucketSelector;
+    private static final Logger log = Logger.getLogger(BucketSelector.class);
+
+    private static final int MAX_PREFIX_LENGTH = 1;
+    private final String rangeMin;
+    private final String rangeMax;
+    private final TreeSet<String> bucketList = new TreeSet<String>();
+
+    public Iterator<String> getBucketIterator() {
+        return bucketList.iterator();
+    }
+
+    public BucketSelector(String selector) {
+        InventoryUtil.assertNotNull(BucketSelector.class, "selectors", selector);
+
+        String[] minMax = selector.split("-");
+        StringBuilder errMsg = new StringBuilder();
+
+        if (minMax.length > 2) {
+            throw new IllegalArgumentException("invalid bucket selector: single value or range only: "
+                + selector);
+        }
+
+        // trim and convert to lower case for consistent processing
+        rangeMin = minMax[0].trim().toLowerCase();
 
-    BucketSelector(String selectorPrefix) {
-        this.bucketSelector = selectorPrefix;
-        throw new UnsupportedOperationException("TODO");
+        if (minMax.length == 1) {
+            rangeMax = rangeMin;
+        } else {
+            rangeMax = minMax[1].trim().toLowerCase();
+        }
+
+        if (!StringUtil.hasLength(rangeMin)) {
+            errMsg.append("empty min range value: " + rangeMin + "\n");
+        }
+        if (!StringUtil.hasLength(rangeMax)) {
+            errMsg.append("empty max range value: " + rangeMax + "\n");
+        }
+
+        if (rangeMin.length() > MAX_PREFIX_LENGTH) {
+            errMsg.append("min range value greater than maxlen (" + MAX_PREFIX_LENGTH + "): " + rangeMin + "\n");
+        }
+        if (rangeMax.length() > MAX_PREFIX_LENGTH) {
+            errMsg.append("max range value greater than maxlen (" + MAX_PREFIX_LENGTH + "): " + rangeMax + "\n");
+        }
+
+        // 0-f is acceptable range
+        int min;
+        int max;
+        min = Artifact.HEXVALUES.indexOf(rangeMin);
```

Comment:
- lines 127 and 131: in the case where rangeMin.length() == 1, Artifact.URI_BUCKET_CHARS.indexOf(String) is OK, but this check for the index of rangeMin isn't generally correct; in general, each character in rangeMin needs to be looked up in the Artifact.URI_BUCKET_CHARS individually to validate rangeMin (and rangeMax)... you can decide if a comment explaining the short cut or a char-by-char validation is needed

Reply:
The algorithm for populating bucketList would need to change as well if MAX_PREFIX_LENGTH > 1.  Because of that, I'm opting for adding a comment for both this and the section where bucketList is populated.